Documentation updates for v12 (#4880)
cypress-io/cypress-documentation: content/guides/guides/cross-origin-testing.md

---
title: Cross Origin Testing
---

<Alert type="info">

<strong class="alert-header"> Note </strong>

As of Cypress [v12.0.0](https://on.cypress.io/changelog#12-0-0), Cypress has the
capability to visit multiple origins in a single test via the
[cy.origin()](https://on.cypress.io/origin) command!

</Alert>

> Cypress limits each test to visiting domains that share the same superdomain. If

Reviewer: > Cypress limits each test to visiting domains that share the same superdomain

This statement feels like it counters the above statement

> As of Cypress [v12.0.0](https://on.cypress.io/changelog#12-0-0), Cypress has the
capability to visit multiple origins in a single test 

I assume you're trying to something like: "by default, cypress runs the commands in a test agains a single origin".
(or is is super domain?  this always confuses me since its cy.origin not cy.superdomain)

Author: it's confusing, but it is superdomain. Once its outside of same superdomain, you have to use `origin()`, which we want an exact match of the said origin and not the origins superdomain. Not exactly easy to digest, but I think what we have gives a pretty good summary.

Author: Without the full context I think it does counter it, but the next sentence ` If
a navigation occurs that does not meet the same superdomain rule, the
[cy.origin](/api/commands/origin) command must be used to execute Cypress commands
inside the newly navigated origin.` I feel like addresses that explanation

a navigation occurs that does not meet the same superdomain rule, the
[`cy.origin`](/api/commands/origin) command must be used to execute assertions
inside the newly navigated origin.

But what is same superdomain? It is actually very similar to that of same
origin! Two URLs have the same origin if the protocol, port (if specified), and

> host are the same for both. Cypress automatically handles hosts of the same

Reviewer: ```suggestion
host match. Cypress automatically handles hosts of the same
```

Reviewer: > Cypress automatically handles hosts of the same

IMO I think you should open this guide with this statement of how cypress works default then go into what is as uper domain and where origin comes into play

Author: I almost feel like you need the opposite in order to understand why cypress works the way it does by default? But I also feel like it takes a minute to get into it 😅

Author: got the suggestion applied in a1c9e8b71

superdomain by injecting the
[`document.domain`](https://developer.mozilla.org/en-US/docs/Web/API/Document/domain)
property into the visited `text/html` pages. This is why navigations without the
use of the [`cy.origin()`](/api/commands/origin) command are solely scope to the
same superdomain.

Given the URLs below, all have the same superdomain compared to
`https://www.cypress.io`.

- `https://cypress.io`
- `https://docs.cypress.io`
- `https://example.cypress.io/commands/querying`

The URLs below, however, will have different superdomains/origins compared to
`https://www.cypress.io`.

- `http://www.cypress.io` (Different protocol)

> - `https://docs.cypress.io:81` (Different port)

Reviewer: what is the default port when one isn't explicitly called out in a url? is there one?

Author: usually `80`

- `https://www.auth0.com/` (Different host of different superdomain)

The `http://localhost` URLs differ if their ports are different. For example,
the `http://localhost:3000` URL is considered to be a different origin from the
`http://localhost:8080` URL.

> The rules are:

Reviewer: Maybe `When to use cy.origin()`?
I thought this was rules of what made superdomain

Author: The rules should still apply even without `cy.origin()`

- <Icon name="exclamation-triangle" color="red"></Icon> You **cannot**
  [visit](/api/commands/visit) two domains of different superdomains in the same
  test and continue to interact with the page without the use of the
  [`cy.origin`](/api/commands/origin) command.
- <Icon name="check-circle" color="green"></Icon> You **can**
  [visit](/api/commands/visit) two or more domains of different origin in
  **different** tests without needing [`cy.origin`](/api/commands/origin).

We understand this is a bit complicated to understand, so we have built a nifty
chart to help clarify the differences!

> ### Parts of a URL

Reviewer: I think it would have been helpful for me to see this diagam before you compared what urls were different superdomains/origins

Author: hoisted this up in a1c9e8b71. Seems to flow better too!

│          │  │           hostname           │ port │ pathname │     search     │       │
|          |  ├───────────┬──────────────────┤      │          │                │       │
│          │  │ subdomain │ superdomain (sd) │      │          │                │       │
|          |  ├───────────┼─────────┬────────┤      │          ├─┬──────────────┤       │
│          │  │           │  domain │  TLD   │      │          │ │    query     │       │
│          │  │           │         │        │      │          │ │              │       │
"  https:   //     sub    . example .  com   : 8080   /p/a/t/h  ?  query=string   #hash "
│                                                   │          │                │       │
│                      origin                       │          |                │       │
├─────────────┬───────────┬─────────────────────────┤          │                │       │
│ (sd) origin │           │   (sd) origin           │          │                │       │
└─────────────┴───────────┴─────────────────────────┴──────────┴────────────────┴───────┘
```

For practical purposes, this means the following:

> ```javascript

Reviewer: This test will run without error:

Author: addressed in a1c9e8b71

it('navigates', () => {
  cy.visit('https://www.cypress.io')
  cy.visit('https://docs.cypress.io')
  cy.get('selector') // yup all good
})
```

> ```javascript

Reviewer: The following test changes super-domain mid-test and results in an error after navigation:

it('navigates', () => {
  cy.visit('https://www.cypress.io')
  cy.visit('https://stackoverflow.com')

>   cy.get('selector') // this will error w/o cy.origin wrap

Reviewer: ```suggestion
  cy.get('selector') // this will error because stackoverflow.com doesn't match the cypress.io  host
```

Author: I wound up moving this comment up to take the place of the previous suggestion to make it easier for users to read inline in a1c9e8b71

})
```

> ```javascript
> it('navigates', () => {
>   cy.visit('https://www.cypress.io')
>   cy.visit('https://stackoverflow.com')
>   cy.origin('https://stackoverflow.com', () => {
>     cy.get('selector') // yup all good
>   })
> })

Reviewer: ```suggestion
To fix the above cross-origin error,  use `cy.origin()` to indicate which origin the sequential command should run against:
```diff
it('navigates', () => {
  cy.visit('https://www.cypress.io')
  cy.visit('https://stackoverflow.com')
+  cy.origin('https://stackoverflow.com', () => {
    cy.get('selector') // yup all good
+  })
})
```

Author: addressed in a1c9e8b71. I didn't implement the diff because the contract looks strange next to the other javascript snippets, but added the explainer

```

> 

Author: I don't think so. I don't think this is a pattern we usually want to encourage, but we want to showcase the exception and let users see if its applicable to their use case or not

> 

Author: I guess not so much discourage but just not common I think

```javascript
it('navigates', () => {
  cy.visit('https://www.cypress.io')
})

// split visiting different origin in another test
it('navigates to new origin', () => {
  cy.visit('https://stackoverflow.com')
  cy.get('selector') // yup all good
})
```

This limitation exists because Cypress switches to the domain under each
specific test when it runs. For more information on this, please see our Web
Security page regarding
[Different superdomain per test requires cy.origin command](/guides/guides/web-security#Different-superdomain-per-test-requires-cy-origin-command).

#### Other workarounds

> There are other ways of testing the interaction between 2 superdomains. Because

Reviewer: ```suggestion
There are other ways of testing the interaction between two superdomains. Because
```

Author: addressed in 9d132e9df

the browser has a natural security barrier called `origin policy` this means
that state like `localStorage`, `cookies`, `service workers` and many other APIs
are not shared between them anyways. Cypress does offer APIs around
`localStorage`, `sessionStorage`, and `cookies` that are not limited to this

> restriction.

Reviewer: With @chrisbreiding's recent changes we can grab & clear local across domains. is this statement true? Also not sure what addin content is trying to say in relation to other multi-domain test strategies

Author: It should be true because of those changes I believe. It's not really to help a testing strategy but to inform the user its available since we've said it isn't in the past and that precident has changed

#### Other workarounds

There are other ways of testing the interaction between two superdomains. The
browser has a natural security barrier called `origin policy` this means that
state like `localStorage`, `cookies`, `service workers` and many other APIs are
not shared between them anyways. Cypress does offer APIs around `localStorage`,
`sessionStorage`, and `cookies` that are not limited to this restriction.

As a best practice, you should not visit or interact with a 3rd party service
not under your control. However, there are exceptions! If your organization uses
Single Sign On (SSO) or OAuth then you might involve a 3rd party service other
than your superdomain, which can be safely tested with
[`cy.origin()`](/api/commands/origin).

We've written several other guides specifically about handling this situation.

- [Best Practices: Visiting external sites](/guides/references/best-practices#Visiting-external-sites)
- [Web Security: Common Workarounds](/guides/guides/web-security#Common-Workarounds)
- [Recipes: Logging In - Single Sign On](/examples/examples/recipes#Logging-In)
- [Guides: Amazon Cognito Authentication](/guides/end-to-end-testing/amazon-cognito-authentication)

> - [Guides: Okta Authentication](/guides/end-to-end-testing/okta-authentication)

Reviewer 2: Can we add Auth0 guide here plz

Author: yep I'll do a follow up to address these comments!
